#1802: Handle empty GZIP streams gracefully

baldram opened the pull request:
Prevents EOFException when processing empty streams in 204 responses and when Content-Length header is absent.

Closes #1802

- [x] Check if the project compiles by running `sbt compile`
- [x] Verify docs compilation by running `sbt compileDocs`
- [x] Check if tests pass by running `sbt test`
- [x] Format code by running `sbt scalafmt`

baldram commented:
Could we also backport this fix to v3 while users prepare for v4 migration?

baldram commented on core/src/main/scalajvm/sttp/client4/internal/SafeGZIPInputStream.scala:
```diff
@@ -0,0 +1,55 @@
+}
+
+/** A safe wrapper for GZIPInputStream that handles empty streams gracefully. This class prevents EOFException when
+  * reading empty or invalid GZIP streams by falling back to a no-op stream.
```
Right, that comment is too far, will remove it! We keep other failures as documented in tests.

baldram replied:
Adjusted.

adamw replied:
It's not just the comment - I think we should only intercept & suppress EOFExceptino if we haven't read any bytes yet. Otherwise, it should be propagated?

baldram replied:
Ok, I will do a closer look regarding byte-counting. The idea so far was that we ensure here unchanged behaviour:
<img width="439" alt="image" src="https://github.com/user-attachments/assets/8da79c63-08e9-4bfd-9098-ca1c899d93d9" />

We just create the `GZIPInputStream` as it happens today. Then if an exception is thrown, we suppress `EOFException` only. Everything else will be thrown as it is today with no change. And this is proved in the test `it should "let callers handle other failures"`. And failing is the same as when using the original Java's GZIPInputStream.

But I will try to think more about concerns that you've raised.

adamw replied:
Yes, sure, we want to keep GZipInputStream intact (delegating all the reading there), it's just that if an `EOFException` happens at any point **after** some data is already returned, then we should propagate it

baldram replied:
> EOFException happens at any point after some data is already returned

Ok, need to think how to prove that, as I think we do not suppress even EOFException in such a scenario, it will be thrown from some other place and we will not prevent it. As the only case is this factory method, when we just create an instance of GZIPInputStream for the first time and it fails immediately then. So this would be before we start reading. But depends how it is later used 🤔 Need to rethink then, hmm.

baldram replied:
>  if an EOFException happens at any point after some data is already returned, then we should propagate it

I managed to add one more test scenario, to test "propagating EOFException during read operations". So it suppress only EOFException when creating a new instance of GZipInputStream and it's fed with an empty stream. When we have a stream that we can start working with it, and later we face EOFException, it will be thrown.

We can observe this behaviour in the new test scenario. 
I wonder if we need to deal additionally with mentioned byte-counting in such a case?

baldram replied:
Ok, added one more test scenario to prove we preserve the original behaviour regarding the end of compressed data. So we do not break any behaviour during the later stream usage. There is still an open point as discussed:
-  current implementation suppresses all EOFExceptions that occur during the creation of the GZIPInputStream
- and we want to be sure that we only suppress EOFExceptions if no data has been read yet
- propagate EOFExceptions if they occur after some data has already been read

I will get back to it later today, probably in the evening, and try to implement a more nuanced approach. The mentioned byte-counting, or so.

adamw replied:
Ah, if the exception that we want to intercept is thrown at construction - great - no need for byte counting :)

adamw commented:
Maybe we could also have a `HttpTest` where we return an empty response with the compression headers set? As I understand, that's the core issue that we're trying to solve here

baldram commented:
> Maybe we could also have a `HttpTest` where we return an empty response with the compression headers set? As I understand, that's the core issue that we're trying to solve here

Yes, that would be the best. I implemented already multiple issue reproducers here: https://gist.github.com/baldram/cfcb9203baaf968107f1d3da44ae1150
But I used Wiremock there. 

And I know, I will not introduce WireMock to the project :) 
I should use existing infrastructure and HttpTest of course.

For some reason I have an issue with Sttp project and existing HttpTests, I can't run them from IntelliJ, they do not work.
Like the server is not started (java.net.ConnectException). I couldn't find in README.md any hints in Contributing section, or I overlooked... Any hints here?

adamw commented:
Ah, running tests from IntelliJ rarely worked for me, so I usually just run them from sbt. The test server is automatically started there.

baldram commented:
> Maybe we could also have a `HttpTest` where ...

@adamw OK. Added a problem reproducer to each backend.
<img width="693" alt="image" src="https://github.com/user-attachments/assets/6e646d62-34ad-4080-8138-b5bed094a197" />

The issue is resolved and confirmed by integration tests. 

While it's challenging to force the Akka-based test server to exclude the Content-Length header for the reported scenario (202 + GZip + no-content + missing Content Length header), the problem was still reproduced. The original issue occurred even with the header present, as there was no length check before proceeding. The implemented fix prevents crashes in both cases: with and without the Content-Length header.

I also adjusted Scaladoc to clarify more a scenario the SafeGZIPInputStream addresses.

At this point, I think the changes are complete. Could you please do a final pass to check if there's anything else that should be addressed? Or we are ready to go?

PS: I'd also like to backport this fix to v3. The code base could changed a bit. Does this require a separate MR, or how can we do this?

baldram commented:
@adamw Ok, all CI is green. Improvements done. Ready for doing a final pass to check if everything is addressed.
Some backends do not use `GZipInputStreamDecompressor` and had to add extra implementation.

At this point, I think the changes are complete.
Are we good to go?
Could we also backport this fix to v3?

adamw commented:
Thanks! :) As for back-porting to v3, could you create a PR against the `v3` branch? I'll merge & release it there then

baldram commented:
> Thanks! :) As for back-porting to v3, could you create a PR against the `v3` branch? I'll merge & release it there then

Ok, and thanks for your support!
